### PR TITLE
Start adding support for TypeScript

### PR DESCRIPTION
**Pull request by keggsmurph21**
[TypeScript](https://www.typescriptlang.org/), if you're not familiar, is basically JavaScript with types.  My thinking is that having types will make refactoring / adding new features _much_ easier, in the long run.

This PR takes some baby steps towards converting the project into TypeScript (luckily, this can be done incrementally).  At the moment, the only thing that has types in `notatrix/src/utils/` (just 5 files).

If you're just working on `client/` or `server/`, this change should be totally transparent.

If you're working on `notatrix/`, then you'll need to make sure you re-compile the TypeScript before rebuilding the server bundle.  In other words, if you're used to

    $ npm run build

then you'll now need to

    $ cd notatrix/
    $ npx tsc
    $ cd ../
    $ npm run build

Eventually, I'd like to remove this^ restriction, but it'd be a lot more boilerplate I think...

<hr>

If you want to take a look, **I'd recommend reviewing commit-by-commit**!

### CODE REVIEW

File: notatrix/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (keggsmurph21) [inline notatrix/tsconfig.json:1]: This config file is documented here: https://www.typescriptlang.org/tsconfig

File: notatrix/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+    "compilerOptions": {
+        "allowJs": true,
```
Comment (keggsmurph21) [inline notatrix/tsconfig.json:3]: Normally, the compiler will yell if you mix `*.ts` and `*.js` files.  Since we want to migrate incrementally, we'll want to enable this flag (for now).

File: notatrix/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+        "outDir": "./build/js",
```
Comment (keggsmurph21) [inline notatrix/tsconfig.json:6]: Where should we dump the generated files?

File: notatrix/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+        "target": "es5"
```
Comment (keggsmurph21) [inline notatrix/tsconfig.json:7]: Which "flavor" of JavaScript should we emit?  This is a shockingly confusing corner of the language, but the basic idea is that the `es5` target is basically universally supported at this point.  I think, honestly, a good direction ultimately would be to change this to `es6`, migrate `client/` + `server/` to TypeScript also, and then get rid of Babel.  I really wish the JavaScript ecosystem were a little more stable, as this is all extremely subtle :cry:

File: notatrix/test/generator.js [hunk 5]
```diff
@@ -103,7 +103,7 @@ describe("generator", () => {
               expect(Array.isArray(generated.loss)).to.equal(true);
 
             } catch (e) {
-              if (e instanceof utils.GeneratorError) {
+              if (e instanceof nx.GeneratorError) {
```
Comment (keggsmurph21) [inline notatrix/test/generator.js:106]: :shrug:

File: notatrix/src/utils/constants.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
+"use strict";
```
Comment (keggsmurph21) [inline notatrix/src/utils/constants.ts:1]: Hm, I guess GitHub thinks this is unrelated to `constants.js`, so it's not showing a side-by-side diff.  The changes here (and in the next few commits) were basically to convert [CommonJS (Node) modules](https://nodejs.org/api/modules.html#moduleexports) to [ES6 modules](https://hacks.mozilla.org/2015/08/es6-in-depth-modules/).  That means

```js
// common.js
exports = {
  a: ...,
  b: ...,
  c: ...,
};
```
becomes
```js
// es6.js
export const a = ...;
export const b = ...;
export const c = ...;
```

File: notatrix/src/utils/errors.ts [hunk 7]
```diff
@@ -1,10 +1,13 @@
...
 
-class NotatrixError extends Error {};
+export class NotatrixError extends Error {}
 
-class ToolError extends NotatrixError {};
+export class ToolError extends NotatrixError {}
+
+export class SplitterError extends ToolError {
+  text: string
+  options: any
```
Comment (keggsmurph21) [inline notatrix/src/utils/errors.ts:9]: annotations!

File: notatrix/src/utils/errors.ts [hunk 8]
```diff
@@ -1,10 +1,13 @@
 "use strict";
 
-class NotatrixError extends Error {};
```
Comment (keggsmurph21) [inline notatrix/src/utils/errors.ts:3]: Semicolons aren't needed after class declarations.

File: notatrix/src/utils/index.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+"use strict";
+
+import * as _ from "underscore";
+
+import * as constants from "./constants";
+import * as errors from "./errors";
+import * as funcs from "./funcs";
+import * as regex from "./regex";
+
+module.exports = _.extend({re: regex}, errors, constants, funcs);
```
Comment (keggsmurph21) [inline notatrix/src/utils/index.ts:10]: This file is a bit weird.. preferably we would write this as
```ts
export * from "./constants";
export * from "./errors";
export * from "./funcs";
export * as re from "./regex";
```
but that would require us to change a bunch of our usage-sites (particularly in the tests).  I think I'm OK just leaving this as-is for now and we can revisit later on when more stuff is migrated.

**Comment by keggsmurph21** [comment]
@jonorthwash @ftyers @Zensho do you think this would be worthwhile?

File: notatrix/tsconfig.json (hunk 4 above)
Comment (jonorthwash) [reply notatrix/tsconfig.json:7]: What do we need that's part of `es6`?

**Comment by jonorthwash** [comment]
I think this is a good idea, from the little I know about TypeScript.  I'm not able to review the commit in any competent way, so if you think it's good we can probably just merge it.

**Comment by jonorthwash** [comment]
Are the new build commands documented somewhere?

**Comment by Zensho** [comment]
I think this looks great and while it might not help with my current backend rewriting that much, it would make any further adjustments and fixes easier once we do a complete typescript transformation of the current code.

File: notatrix/tsconfig.json (hunk 4 above)
Comment (keggsmurph21) [reply notatrix/tsconfig.json:7]: IIRC, there's some weirdness in ES5 around extending certain builtin classes like `Array` or `Error`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#es5_customerror_object

**Comment by keggsmurph21** [comment]
> Are the new build commands documented somewhere?

Haha, great question! I added some documentation in https://github.com/jonorthwash/ud-annotatrix/pull/496/commits/a1da42a44e4ba9ad23606c9ebc29ba857a84f367.  Let me know if there's anything else you want me to include.

> I'm not able to review the commit in any competent way, so if you think it's good we can probably just merge it.

Ok, sounds good! I just wanted to make sure you had a chance to veto before I hit "merge" :grin:
